### PR TITLE
[NativeAOT-LLVM] Enhance dumps, enable LIR checking, implement `FLOAT` returns

### PR DESCRIPTION
**Pull request by SingleAccretion**
Also delete a bit of dead code, rename `lowerToShadowStack` and use the Jit's own hash tables.

The main point is the nice IR dumping we will now get from codegen:
```cs
[MethodImpl(MethodImplOptions.NoInlining)]
private static float Problem(StructWithIndex* a)
{
    for (int i = 0; i < a->Index; i++)
    {
        a->Value++;
    }

    return a->Value;
}

struct StructWithIndex
{
    public int Index;
    public int Value;
}
```
```scala
*************** Starting PHASE Build LLVM

=============== Generating BB01 [000..004) -> BB03 (cond), preds={} succs={BB04,BB03} flags=0x00000000.20000020: i LIR
Generating:                [000061] ------------                 IL_OFFSET void   IL offset: 0x0
Generating: N001 (  1,  1) [000000] ------------         t0 =    CNS_INT   int    0 $40
                                                              /--*  t0     int
Generating: N003 (  1,  3) [000002] DA----------              *  STORE_LCL_VAR int    V01 loc0         d:2
Generating: N001 (  1,  1) [000042] ------------        t42 =    LCL_VAR   int    V00 arg0         u:1 $80
                                                              /--*  t42    int
Generating: N002 (  3,  2) [000041] ---XG-------        t41 = *  IND       int    <l:$181, c:$180>
  %2 = bitcast %"[RyuJitReproduction]RyuJitReproduction.Program+StructWithIndex"* %1 to i32*
  %3 = load i32, i32* %2, align 4
Generating: N003 (  1,  1) [000060] ------------        t60 =    CNS_INT   int    0 $40
                                                              /--*  t41    int
                                                              +--*  t60    int
Generating: N004 (  5,  4) [000039] J--XG--N----        t39 = *  LE        int    <l:$185, c:$184>
  %4 = icmp sle i32 %3, 0
                                                              /--*  t39    int
Generating: N005 (  7,  6) [000038] ---XG-------              *  JTRUE     void
  br i1 %4, label %BB03, label %BB04

=============== Generating BB04 [004..???), preds={BB01} succs={BB02} flags=0x00000000.24000440: internal nullcheck LoopPH LIR
Generating: N001 (  1,  1) [000051] -------H----        t51 =    LCL_VAR   int    V00 arg0         u:1 $80
Generating: N002 (  1,  1) [000052] -------H----        t52 =    CNS_INT   int    4 field offset Fseq[Value] $43
                                                              /--*  t51    int
                                                              +--*  t52    int
Generating: N003 (  3,  3) [000050] -------H----        t50 = *  ADD       int    $187
  %5 = bitcast %"[RyuJitReproduction]RyuJitReproduction.Program+StructWithIndex"* %1 to i8*
  %6 = getelementptr i8, i8* %5, i32 4
                                                              /--*  t50    int
Generating: N005 (  3,  3) [000056] DA----------              *  STORE_LCL_VAR int    V03 cse0         d:2
  %7 = ptrtoint i8* %6 to i32

=============== Generating BB02 [004..01C) -> BB02 (cond), preds={BB02,BB04} succs={BB03,BB02} flags=0x00000008.21006420: i Loop Loop0 nullcheck bwd bwd-target LIR
Generating:      (  0,  0) [000066] ------------        t66 =    PHI_ARG   int    V01 loc0         u:4
Generating:      (  0,  0) [000065] ------------        t65 =    PHI_ARG   int    V01 loc0         u:2
                                                              /--*  t66    int
                                                              +--*  t65    int
Generating:                [000063] ------------        t63 = *  PHI       int
  %8 = phi i32
                                                              /--*  t63    int
Generating:                [000064] DA----------              *  STORE_LCL_VAR int    V01 loc0         d:3
Generating: N001 (  1,  1) [000059] ------------        t59 =    LCL_VAR   int    V03 cse0         u:2 $187
                                                              /--*  t59    int
Generating: N003 (  1,  3) [000012] DA--G-------              *  STORE_LCL_VAR int    V02 tmp0         d:2
Generating: N001 (  1,  1) [000014] ------------        t14 =    LCL_VAR   int    V02 tmp0         u:2 $187
                                                              /--*  t14    int
Generating: N002 (  3,  2) [000015] *--XG-------        t15 = *  IND       int    <l:$189, c:$188>
  %9 = inttoptr i32 %7 to i32*
  %10 = load i32, i32* %9, align 4
Generating: N003 (  1,  1) [000016] ------------        t16 =    CNS_INT   int    1 $45
                                                              /--*  t15    int
                                                              +--*  t16    int
Generating: N004 (  5,  4) [000017] ---XG-------        t17 = *  ADD       int    <l:$18d, c:$18c>
  %11 = add i32 %10, 1
Generating: N005 (  1,  1) [000013] ------------        t13 =    LCL_VAR   int    V02 tmp0         u:2 (last use) $187
                                                              /--*  t13    int
                                                              +--*  t17    int
Generating:                [000062] -A-XG-------              *  STOREIND  int
  %12 = inttoptr i32 %7 to i32*
  store i32 %11, i32* %12, align 4
Generating: N001 (  1,  1) [000020] ------------        t20 =    LCL_VAR   int    V01 loc0         u:3 (last use) $101
Generating: N002 (  1,  1) [000021] ------------        t21 =    CNS_INT   int    1 $45
                                                              /--*  t20    int
                                                              +--*  t21    int
Generating: N003 (  3,  3) [000022] ------------        t22 = *  ADD       int    $18e
  %13 = add i32 %8, 1
                                                              /--*  t22    int
Generating: N005 (  3,  3) [000024] DA----------              *  STORE_LCL_VAR int    V01 loc0         d:4
Generating: N001 (  1,  1) [000003] ------------         t3 =    LCL_VAR   int    V01 loc0         u:4 $18e
Generating: N002 (  1,  1) [000004] ------------         t4 =    LCL_VAR   int    V00 arg0         u:1 Zero Fseq[Index] $80
                                                              /--*  t4     int
Generating: N003 (  3,  2) [000005] ---XG-------         t5 = *  IND       int    <l:$190, c:$18f>
  %14 = bitcast %"[RyuJitReproduction]RyuJitReproduction.Program+StructWithIndex"* %1 to i32*
  %15 = load i32, i32* %14, align 4
                                                              /--*  t3     int
                                                              +--*  t5     int
Generating: N004 (  5,  4) [000006] J--XG--N----         t6 = *  LT        int    <l:$194, c:$193>
  %16 = icmp slt i32 %13, %15
                                                              /--*  t6     int
Generating: N005 (  7,  6) [000007] ---XG-------              *  JTRUE     void
  br i1 %16, label %BB02, label %BB03

=============== Generating BB03 [01C..024) (return), preds={BB01,BB02} succs={} flags=0x00000000.20000020: i LIR
Generating: N001 (  1,  1) [000025] ------------        t25 =    LCL_VAR   int    V00 arg0         u:1 (last use) $80
Generating: N002 (  1,  1) [000036] ------------        t36 =    CNS_INT   int    4 field offset Fseq[Value] $43
                                                              /--*  t25    int
                                                              +--*  t36    int
Generating: N003 (  2,  2) [000037] -------N----        t37 = *  ADD       int    $187
  %8 = bitcast %"[RyuJitReproduction]RyuJitReproduction.Program+StructWithIndex"* %1 to i8*
  %9 = getelementptr i8, i8* %8, i32 4
                                                              /--*  t37    int
Generating: N004 (  4,  4) [000026] ---XG-------        t26 = *  IND       int    <l:$196, c:$195>
  %10 = bitcast i8* %9 to i32*
  %11 = load i32, i32* %10, align 4
                                                              /--*  t26    int
Generating: N005 ( 10, 10) [000027] ---XG-------        t27 = *  CAST      float <- int <l:$203, c:$202>
  %12 = sitofp i32 %11 to float
                                                              /--*  t27    float
Generating: N006 ( 11, 11) [000028] ---XG-------              *  RETURN    float  $240
  ret float %12

===================================================================================================================
LLVM IR for RyuJitReproduction.Program:Problem(int):float after codegen:
-------------------------------------------------------------------------------------------------------------------

define float @RyuJitReproduction_RyuJitReproduction_Program__Problem(i8* %0, %"[RyuJitReproduction]RyuJitReproduction.Program+StructWithIndex"* %1) {
Prolog:
  br label %BB01

BB01:                                             ; preds = %Prolog
  %2 = bitcast %"[RyuJitReproduction]RyuJitReproduction.Program+StructWithIndex"* %1 to i32*
  %3 = load i32, i32* %2, align 4
  %4 = icmp sle i32 %3, 0
  br i1 %4, label %BB03, label %BB04

BB04:                                             ; preds = %BB01
  %5 = bitcast %"[RyuJitReproduction]RyuJitReproduction.Program+StructWithIndex"* %1 to i8*
  %6 = getelementptr i8, i8* %5, i32 4
  %7 = ptrtoint i8* %6 to i32
  br label %BB02

BB03:                                             ; preds = %BB02, %BB01
  %8 = bitcast %"[RyuJitReproduction]RyuJitReproduction.Program+StructWithIndex"* %1 to i8*
  %9 = getelementptr i8, i8* %8, i32 4
  %10 = bitcast i8* %9 to i32*
  %11 = load i32, i32* %10, align 4
  %12 = sitofp i32 %11 to float
  ret float %12

BB02:                                             ; preds = %BB02, %BB04
  %13 = phi i32 [ %18, %BB02 ], [ 0, %BB04 ]
  %14 = inttoptr i32 %7 to i32*
  %15 = load i32, i32* %14, align 4
  %16 = add i32 %15, 1
  %17 = inttoptr i32 %7 to i32*
  store i32 %16, i32* %17, align 4
  %18 = add i32 %13, 1
  %19 = bitcast %"[RyuJitReproduction]RyuJitReproduction.Program+StructWithIndex"* %1 to i32*
  %20 = load i32, i32* %19, align 4
  %21 = icmp slt i32 %18, %20
  br i1 %21, label %BB02, label %BB03
}


*************** Finishing PHASE Build LLVM
```

cc @yowl

### CODE REVIEW

File: src/coreclr/jit/llvm.cpp [hunk 1]
```diff
@@ -92,12 +92,16 @@ Llvm::Llvm(Compiler* compiler)
     : _compiler(compiler),
     _info(compiler->info),
     _function(nullptr),
+    _sigInfo(compiler->info.compMethodInfo->args),
     _builder(_llvmContext),
     _prologBuilder(_llvmContext),
+    _blkToLlvmBlkVectorMap(compiler->getAllocator(CMK_Codegen)),
+    _sdsuMap(compiler->getAllocator(CMK_Codegen)),
+    _localsMap(compiler->getAllocator(CMK_Codegen)),
+    _debugMetadataMap(compiler->getAllocator(CMK_Codegen)),
```
Comment (SingleAccretion) [inline src/coreclr/jit/llvm.cpp:101]: > `_debugMetadataMap`

@yowl I wanted to ask - what is the point of this map? We use it in `getOrCreateDebugMetadata`, but only once per compilation, so it should always go the "lookup failed" path (of which I am not sure what the consequences are).

Comment (yowl) [reply src/coreclr/jit/llvm.cpp:101]: I think I was basically just copying

https://github.com/dotnet/runtimelab/blob/9c53250461c337d6febcdc2f43ea9619123b8354/src/coreclr/tools/aot/ILCompiler.LLVM/Compiler/LLVMCodegenCompilation.cs#L59

Where it asks for the debug compilation unit for every IL instruction.

**Comment by yowl** [comment]
This is really neat the way you've merged the dumps, Looks really good.

**Comment by jkotas** [comment]
Is this ready to merge, or do you plan to incorporate follow up on https://github.com/dotnet/runtimelab/pull/2082#discussion_r1020958397 ?

**Comment by SingleAccretion** [comment]
@jkotas Ready, will investigate the debug info question separately.

**Pull request merged**